### ISSUE TITLE
Model.test_on_batch reset_metrics incorrect

### ISSUE THREAD

**Issue by jackd** [tensorflow/tensorflow#45883]
URL with issue: [Model.test_on_batch](https://www.tensorflow.org/api_docs/python/tf/keras/Model#test_on_batch)

## Description of issue (what needs changing):
`reset_metrics` states: 

> If True, the metrics returned will be only for this batch. If False, the metrics will be statefully accumulated across batches.

This is incorrect, as metric states are only reset AFTER values are computed. See example below.

This could be considered a bug, but it might be considered a breaking change to fix it, and addressing it as a documentation issue may be preferable.

```python
import tensorflow as tf

class Counter(tf.keras.metrics.Metric):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.count = self.add_weight('count', dtype=tf.int64, initializer='zeros')

    def update_state(self, *args, **kwargs):
        self.count.assign_add(1)

    def result(self):
        return self.count


tf.random.set_seed(0)
inp = tf.keras.Input((1,))
out = tf.keras.layers.Dense(2, activation='softmax')(inp)
model = tf.keras.Model(inp, out)

model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(), metrics=[Counter()])
batch_size = 1
x = tf.zeros((batch_size, 1,), dtype=tf.float32)
y = tf.zeros((batch_size,), dtype=tf.int64)
model.train_on_batch(x, y, reset_metrics=False)
logs = model.test_on_batch(x, y, reset_metrics=True, return_dict=True)
#  logs contains metrics for both both steps
print(f"counter = {logs['counter']}")  # counter = 2
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/61d5f07d5daa4ed4355b9383b8e8f03f/45883.ipynb#scrollTo=fLJhsUW1Swlj). Thanks!

**Comment by rmothukuru** [comment]
@jackd,
Can you please elaborate on what the Expected Behavior should be? Thanks!

**Comment by jackd** [comment]
Based on the documentation, I would expect counter to always be `1` when `return_dict=True` (i.e. `reset_metrics` should be called before metrics are updated and results are computed, rather than afterwards).

**Comment by rmothukuru** [comment]
@jackd,
Would you be interested to submit a pull request to fix the issue? 

If so, please see the docs
contributor guide: https://www.tensorflow.org/community/contribute/docs,
docs API guide: https://www.tensorflow.org/community/contribute/docs_ref and the
docs style guide: https://www.tensorflow.org/community/contribute/docs_style

**Comment by jackd** [comment]
@rmothukuru sorry, work's been busy. Are you looking for a bug fix or just making the documentation consistent with the current behaviour?

**Comment by sachinprasadhs** [comment]
I'm able to to see the expected output, `counter=1`, [here](https://colab.research.google.com/gist/sachinprasadhs/a18cb67463c00482aa447a546fff331f/45883.ipynb) is the gist for the same.
Please move this issue to close if your issue is resolved. Thanks!

**Comment by jackd** [comment]
Looks to work as of 2.7, thanks.

**Comment by sachinprasadhs** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
